SCons/Util.py

```diff
@@ -525,71 +612,65 @@ def to_String_for_signature(obj, to_String_for_subst=to_String_for_subst,
-        if key not in exclude:
-            copy[key] = semi_deepcopy(val)
-    return copy
-d[dict] = semi_deepcopy_dict
-
-def _semi_deepcopy_list(x):
-    return list(map(semi_deepcopy, x))
-d[list] = _semi_deepcopy_list
-
-def _semi_deepcopy_tuple(x):
-    return tuple(map(semi_deepcopy, x))
-d[tuple] = _semi_deepcopy_tuple
-
-def semi_deepcopy(x):
-    copier = _semi_deepcopy_dispatch.get(type(x))
+def semi_deepcopy_dict(obj, exclude=None) -> dict:
+    if exclude is None:
+        exclude = []
+    return {k: semi_deepcopy(v) for k, v in obj.items() if k not in exclude}
+
+def _semi_deepcopy_list(obj) -> list:
+    return [semi_deepcopy(item) for item in obj]
+
+def _semi_deepcopy_tuple(obj) -> tuple:
+    return tuple(map(semi_deepcopy, obj))
+
+_semi_deepcopy_dispatch = {
+    dict: semi_deepcopy_dict,
+    list: _semi_deepcopy_list,
+    tuple: _semi_deepcopy_tuple,
+}
+
+def semi_deepcopy(obj):
+    copier = _semi_deepcopy_dispatch.get(type(obj))
     if copier:
-        return copier(x)
-    else:
-        if hasattr(x, '__semi_deepcopy__') and callable(x.__semi_deepcopy__):
-            return x.__semi_deepcopy__()
-        elif isinstance(x, UserDict):
-            return x.__class__(semi_deepcopy_dict(x))
-        elif isinstance(x, UserList):
-            return x.__class__(_semi_deepcopy_list(x))
+        return copier(obj)
 
-        return x
+    if hasattr(obj, '__semi_deepcopy__') and callable(obj.__semi_deepcopy__):
+        return obj.__semi_deepcopy__()
+
+    if isinstance(obj, UserDict):
+        return obj.__class__(semi_deepcopy_dict(obj))
+
+    if isinstance(obj, UserList):
+        return obj.__class__(_semi_deepcopy_list(obj))
+
+    return obj
 
 
 class Proxy:
-    """A simple generic Proxy class, forwarding all calls to
-    subject.  So, for the benefit of the python newbie, what does
-    this really mean?  Well, it means that you can take an object, let's
-    call it 'objA', and wrap it in this Proxy class, with a statement
-    like this
+    """A simple generic Proxy class, forwarding all calls to subject.
+
+    This means you can take an object, let's call it `'obj_a`,
+    and wrap it in this Proxy class, with a statement like this::
 
-                 proxyObj = Proxy(objA),
+        proxy_obj = Proxy(obj_a)
 
-    Then, if in the future, you do something like this
+    Then, if in the future, you do something like this::
 
-                 x = proxyObj.var1,
+        x = proxy_obj.var1
 
-    since Proxy does not have a 'var1' attribute (but presumably objA does),
-    the request actually is equivalent to saying
+    since the :class:`Proxy` class does not have a :attr:`var1` attribute
+    (but presumably `objA` does), the request actually is equivalent to saying::
 
-                 x = objA.var1
+        x = obj_a.var1
 
     Inherit from this class to create a Proxy.
 
     Note that, with new-style classes, this does *not* work transparently
```

Comment:
This could probably be changed since now all classes are new style.. ;)

Reply:
Maybe just dump a TODO: here noting that?

Reply:
Indeed - a TODO sounds useful.